libcxx/test/std/utilities/meta/meta.unary/meta.unary.prop/is_implicit_lifetime.pass.cpp

```diff
@@ -0,0 +1,271 @@
+    requires B
+  = default;
+  ConstrainedUserDeclaredDefaultConstructor(const ConstrainedUserDeclaredDefaultConstructor&) {}
+};
+
+template <bool B>
+struct ConstrainedUserProvidedDestructor {
+  ~ConstrainedUserProvidedDestructor() = default;
+  ~ConstrainedUserProvidedDestructor()
+    requires B
+  {}
+};
+
+struct StructWithFlexibleArrayMember {
+  int arr[];
+};
+
+struct StructWithZeroSizedArray {
+  int arr[0];
+};
+
+// Test implicit-lifetime type
+template <typename T, bool Expected>
+constexpr void test_is_implicit_lifetime() {
+  assert(std::is_implicit_lifetime<T>::value == Expected);
+  assert(std::is_implicit_lifetime_v<T> == Expected);
+}
+
+// Test pointer, reference, array, etc. types
+template <typename T>
+constexpr void test_is_implicit_lifetime() {
+  test_is_implicit_lifetime<T, true>();
+
+  // cv-qualified
+  test_is_implicit_lifetime<const T, true>();
+  test_is_implicit_lifetime<volatile T, true>();
+
+  test_is_implicit_lifetime<T&, false>();
+  test_is_implicit_lifetime<T&&, false>();
+
+  // Pointer types
+  test_is_implicit_lifetime<T*, true>();
+
+  // Arrays
+  test_is_implicit_lifetime<T[], true>();
+  test_is_implicit_lifetime<T[94], true>();
+}
+
+constexpr bool test() {
+  // Standard fundamental C++ types
+
+  test_is_implicit_lifetime<std::nullptr_t, true>();
+
+  test_is_implicit_lifetime<void, false>();
+  test_is_implicit_lifetime<const void, false>();
+  test_is_implicit_lifetime<volatile void, false>();
+
+  test_is_implicit_lifetime<bool>();
+
+  test_is_implicit_lifetime<char>();
+  test_is_implicit_lifetime<signed char>();
+  test_is_implicit_lifetime<unsigned char>();
+
+#if !defined(TEST_HAS_NO_WIDE_CHARACTERS)
+  test_is_implicit_lifetime<wchar_t>();
+#endif
+
+#if !defined(TEST_HAS_NO_CHAR8_T)
+  test_is_implicit_lifetime<char8_t>();
+#endif
+  test_is_implicit_lifetime<char16_t>();
+  test_is_implicit_lifetime<char32_t>();
+
+  test_is_implicit_lifetime<short>();
+  test_is_implicit_lifetime<short int>();
+  test_is_implicit_lifetime<signed short>();
+  test_is_implicit_lifetime<signed short int>();
+  test_is_implicit_lifetime<unsigned short>();
+  test_is_implicit_lifetime<unsigned short int>();
+  test_is_implicit_lifetime<int>();
+  test_is_implicit_lifetime<signed>();
+  test_is_implicit_lifetime<signed int>();
+  test_is_implicit_lifetime<unsigned>();
+  test_is_implicit_lifetime<unsigned int>();
+  test_is_implicit_lifetime<long>();
+  test_is_implicit_lifetime<long int>();
+  test_is_implicit_lifetime<signed long>();
+  test_is_implicit_lifetime<signed long int>();
+  test_is_implicit_lifetime<unsigned long>();
+  test_is_implicit_lifetime<unsigned long int>();
+  test_is_implicit_lifetime<long long>();
+  test_is_implicit_lifetime<long long int>();
+  test_is_implicit_lifetime<signed long long>();
+  test_is_implicit_lifetime<signed long long int>();
+  test_is_implicit_lifetime<unsigned long long>();
+  test_is_implicit_lifetime<unsigned long long int>();
```

Comment:
Here are also a bunch of duplicates. `long`, `long int`, `signed long` and `signed long int` are the same type. Maybe you should just use `arithmetic_types`. That would also remove the need for the `#ifdef`s above.